ggshield/oauth.py

```diff
@@ -0,0 +1,273 @@
+            token=self._access_token,  # type: ignore
+            expire_at=api_token_data.get("expire_at"),
+            token_name=api_token_data.get("name", ""),
+            type=api_token_data.get("type", ""),
+        )
+        instance_config = self.config.auth_config.instances[self.instance]
+        instance_config.account = account_config
+        self.config.save()
+
+    @property
+    def redirect_uri(self) -> str:
+        return f"http://localhost:{REDIRECT_PORT}"
+
+    @property
+    def dashboard_url(self) -> str:
+        return self.config.dashboard_url
+
+    @property
+    def api_url(self) -> str:
+        return self.config.api_url
+
+
+class RequestHandlerWrapper:
+    """
+    Utilitary class to link the server and the request handler.
+    This allows to kill the server from the request processing.
+    """
+
+    oauth_client: OAuthClient
+    complete: bool
+    error_message: Optional[str] = None
+
+    def __init__(self, oauth_client: OAuthClient) -> None:
+        self.oauth_client = oauth_client
+        self.complete = False
+        self.error_message = None
+
+    @property
+    def request_handler(self) -> Type[BaseHTTPRequestHandler]:
+        class RequestHandler(BaseHTTPRequestHandler):
+            def do_GET(self_) -> None:
+                """
+                Handle the get request only.
+                If an authorization code can be extracted from the uri, attach it to the handler
+                so it can be retrieved after the request is processed, then kill the server.
+                """
+                callback_url: str = self_.path
+                parsed_url = urlparse.urlparse(callback_url)
+                if parsed_url.path == "/":
+                    try:
+                        self.oauth_client.process_callback(callback_url)
+                    except OAuthError as error:
+                        self_._end_request()
+                        self.error_message = error.message
+                    else:
+                        self_._end_request(
+                            urlparse.urljoin(
+                                self.oauth_client.dashboard_url, "authenticated"
+                            ),
+                        )
+
+                    # indicate to the serve to stop
+                    self.complete = True
+
+            def _end_request(self_, redirect_url: Optional[str] = None) -> None:
+                """
+                End the current request. If a redirect url is provided,
+                the response will be a redirection to this url.
+                If not the response will be a user error 400
+                """
+                status_code = 301 if redirect_url is not None else 400
+                self_.send_response(status_code)
+
+                if redirect_url is not None:
+                    self_.send_header("Location", redirect_url)
+                self_.end_headers()
+
+        return RequestHandler
```

Comment:
:question: I'm not very experienced with this server handling, but I have a question: why couldn't we just open the server in the main thread? killing the main process would kill the server as we want, and it would make the code a lot simpler to read and understand.
Right now, the code in the OAuth just contains half of the flow.